aips/staking-contract-to-delegation-pool.md

```diff
@@ -0,0 +1,107 @@
+
+Partial voting is automatically enabled. Previously, the staker could set the delegated voter of the underlying stake pool, implicitly of their owned stake. Now, the staker can vote and delegate their voting power as a regular delegator.
+
+If the beneficiary of operator is not configured on `aptos_framework::delegation_pool`, then it cannot be initialized as the conversion function doesn't have access to operator's signer, only to staker's. The default beneficiary will be used which is the operator themselves.
+If the beneficiary is already set, then this address will be used for the new delegation pool as well.
+
+In terms of rewards, there is virtually no impact on the staker, they continue to be rewarded at the same rate as before. In case of the operator, they may expect higher rewards as additional delegators join the pool.
+
+Staker's delegated voter (stake pool's voter) is preserved and applies immediately. However, delegating voting power will apply from the next lockup cycle from now on.
+
+To summarize, state that will be preserved:
+- pool ownership
+- operator
+- commission fee
+- delegated voter of staker
+- individual stakes of staker, operator and any previous operators still owning pending-inactive commission
+
+While, state that will change:
+- the voter of the stake pool is set to the resource account of the delegation pool
+- beneficiary of operator: address already applying on delegation-pool module will be used
+
+## Alternative solutions
+
+As described above, manually converting to a delegation pool requires:
+- doubling the infrastructure for a brief period
+- rewards downtime
+- decreased network security for a brief period
+
+## Specification
+
+`aptos_framework::delegation_pool` implements `initialize_delegation_pool_from_staking_contract(staker: &signer, operator: address)` which is called by `staker` on their staking contract identified by `operator` to atomically convert it, preserving the state of the underlying stake pool, implicitly the activeness of the validator.
+
+`aptos_framework::staking_contract` implements `destroy_staking_contract(staker: &signer, operator: address)` which will only be called by `aptos_framework::delegation_pool::initialize_delegation_pool_from_staking_contract`. This function destroys the `StakingContract` between `staker` and `operator` and returns the resource account storing the stake pool and the ownership capability.
+
+## Reference Implementation
+
+There is a reference implementation at https://github.com/bwarelabs/aptos-core/tree/convert-staking-contract-to-delegation-pool
+
+## Testing (Optional)
+
+Design UTs to validate that:
+- there is no remaining state of the staking contract while the underlying stake pool remains untouched
+- the post-conversion state is consistent
+- the resulted delegation pool accepts delegations, staker is treated as a regular delegator and the commission is correctly distributed
+
+## Risks and Drawbacks
```

Comment:
Is it possible that `initialize_delegation_pool_from_staking_contract` succeeded, but the conversion doesn't happen properly, resulting in the converted delegation pool that does not work as expected and ends up in a weird state?

If that scenario were to happen, would there be a way to manually fix it?

Reply:
The state of the staking contract to be migrated to a newly created delegation pool classifies into:
- ownership of *staker*/*operator*/*previous operators* over the APT locked on the stake pool in the states: *pending-active*, *active*, *pending-inactive* and *inactive*
- pool ownership over both delegation and stake pools
- pool configs: *operator*, *commission*, *beneficiary*, *voter*
- compatibility to `delegation_pool` module:
    * `StakePool`, `DelegationPool` and `stake::OwnerCapability` should be stored on the same resource account
    * resource account to be registered to `AptosCoin`
    * `operator_commission_percentage` in a 2-decimals format

Fortunately, some state variables do not have the drop capability nor an explicit destructor which means they cannot be simply lost: either they remain stored on the staking contract or they are migrated to the delegation pool. If the migration would be somehow not fully consistent, this may allow future patching and fixing.

Ultimately, the correctness of the migration should be validated through unit-testing and formal verification.